## Types of Encoding

In [ ]:
#Nominl Encoding -> Not to worry about Arrangements,(ex- Gender -> Male, Female)
#Ordinal Encoding -> worry about Rank or Arrangements, (Education -> B.E, B.com,p,PhD,Masters = 2,1,4,3)

In [ ]:
#1. Nominal Encoding
##1.1 One Hot Encoding
##1.2 One hot Encoding with many categorical 
##1.3 Mean Encoding

#2. Ordinal Encoding
##2.1 Label Encoding
##2.2 Target Guided ordinal Encoding

### 1.1 One Hot Encoding

In [ ]:
#If I want to apply One Hot Encoding, then how many numbers of categories available that many numbers of columns get created.
####-----------------------------------------------------------------------------------------------------------
#we can do this either with Pandas or sklearn also
#we have converted this category into dummy variables, below are nothing but that is dummy variables
                    Dummy Variables Trap
-------------------------------------------------------------------------------------
#State 		|Germany		India		England
#Germany	|   1               0           0
#India   	|   0               1           0
#England 	|   0               0           1 

#Always Remember, when we do Hot Encoding, Make sure, we have to remove one column either 1st, 2nd or last column anything.
#Suppose, I have 5 categories, then total no of columns get created is (5-1 = 4)


### Disadvantages of One Hot Encoding

In [ ]:
#Considered, If I having pincode column and there is 100 rows, so in that case, 99 columns will be created.
##If the number of columns is Increasing, It means, its increasing the Number of dimensions and that leads to something called Curse of Dimensionality
###So, If you have many no's of category, so don't apply One Hot Encoding.
####So, How can we fix this particular problem ?
#####Let's see in Next Step.


### 1.2 One Hot Encoding with Multiple Categories

In [33]:
#This techniques is basically used in some competition that is KDD orange competition.
#Mainly they found out which is most number of repeating categories and from that, top 10 repeatiting categories is more.
#So,they only applied One Hot Encoding only to the Top 10 categories.

#Suppose, I have 50 categories in One column then we need to apply One hot Encoding to Top 10 columns, then total (10-1=9)columns getting created, Instead of (50-1=49)columns.


## One Hot Encoding - variables with many categories

In [1]:
import pandas as pd
import numpy as np

In [4]:
mercedes_benz = pd.read_csv('Mercedes_train.csv')
mercedes_benz

ID       y  X0 X1  X2 X3 X4  X5 X6 X8  ...  X375  X376  X377  X378  \
0        0  130.81   k  v  at  a  d   u  j  o  ...     0     0     1     0   
1        6   88.53   k  t  av  e  d   y  l  o  ...     1     0     0     0   
2        7   76.26  az  w   n  c  d   x  j  x  ...     0     0     0     0   
3        9   80.62  az  t   n  f  d   x  l  e  ...     0     0     0     0   
4       13   78.02  az  v   n  f  d   h  d  n  ...     0     0     0     0   
...    ...     ...  .. ..  .. .. ..  .. .. ..  ...   ...   ...   ...   ...   
4204  8405  107.39  ak  s  as  c  d  aa  d  q  ...     1     0     0     0   
4205  8406  108.77   j  o   t  d  d  aa  h  h  ...     0     1     0     0   
4206  8412  109.22  ak  v   r  a  d  aa  g  e  ...     0     0     1     0   
4207  8415   87.48  al  r   e  f  d  aa  l  u  ...     0     0     0     0   
4208  8417  110.85   z  r  ae  c  d  aa  g  w  ...     1     0     0     0   

      X379  X380  X382  X383  X384  X385  
0        0     0     0     0     0     0  
1        0     0     0     0     0     0  
2        0     0     1     0     0     0  
3        0     0     0     0     0     0  
4        0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...  
4204     0     0     0     0     0     0  
4205     0     0     0     0     0     0  
4206     0     0     0     0     0     0  
4207     0     0     0     0     0     0  
4208     0     0     0     0     0     0  

[4209 rows x 378 columns]

In [13]:
# we take 6 columns only with the help of usecols
mercedes_benz_col = pd.read_csv('Mercedes_train.csv', usecols=['X0','X1','X2','X3','X4','X5'])
#mercedes_benz_col.head()
mercedes_benz_col.columns

Index(['X0', 'X1', 'X2', 'X3', 'X4', 'X5'], dtype='object')

In [14]:
#Lets took how many labels each variable has
for col in mercedes_benz_col.columns:
    print(col,':',len(mercedes_benz_col[col].unique()),'labels')

X0 : 47 labels
X1 : 27 labels
X2 : 44 labels
X3 : 7 labels
X4 : 4 labels
X5 : 29 labels


In [15]:
# Lets examine how many columns we will obtain after one hot encoding these variables
pd.get_dummies(mercedes_benz_col,drop_first=True).shape

(4209, 152)

##### we can see that from just 6 Initial categorical variables, we end up with 152 new variables
##### what can we do Instead?

In [16]:
# Lets find top 10 most frequent categories for variable X2
mercedes_benz_col.X2.value_counts().sort_values(ascending=False).head(20)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
i       25
k       25
b       21
ao      20
z       19
ag      19
Name: X2, dtype: int64

In [21]:
#Lets make a list with the most frequent categories of the variable
top_10 = [x for x in mercedes_benz_col.X2.value_counts().sort_values(ascending=False).head(10).index]
top_10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [24]:
#And now we make the 10 binary variables

for label in top_10:
    mercedes_benz_col[label] = np.where(mercedes_benz_col['X2']==label,1,0)
mercedes_benz_col[['X2']+top_10].head(40)

X2  as  ae  ai  m  ak  r  n  s  f  e
0   at   0   0   0  0   0  0  0  0  0  0
1   av   0   0   0  0   0  0  0  0  0  0
2    n   0   0   0  0   0  0  1  0  0  0
3    n   0   0   0  0   0  0  1  0  0  0
4    n   0   0   0  0   0  0  1  0  0  0
5    e   0   0   0  0   0  0  0  0  0  1
6    e   0   0   0  0   0  0  0  0  0  1
7   as   1   0   0  0   0  0  0  0  0  0
8   as   1   0   0  0   0  0  0  0  0  0
9   aq   0   0   0  0   0  0  0  0  0  0
10   r   0   0   0  0   0  1  0  0  0  0
11   e   0   0   0  0   0  0  0  0  0  1
12  ai   0   0   1  0   0  0  0  0  0  0
13   e   0   0   0  0   0  0  0  0  0  1
14  as   1   0   0  0   0  0  0  0  0  0
15  ak   0   0   0  0   1  0  0  0  0  0
16   e   0   0   0  0   0  0  0  0  0  1
17  as   1   0   0  0   0  0  0  0  0  0
18   m   0   0   0  1   0  0  0  0  0  0
19  ak   0   0   0  0   1  0  0  0  0  0
20   m   0   0   0  1   0  0  0  0  0  0
21  as   1   0   0  0   0  0  0  0  0  0
22   a   0   0   0  0   0  0  0  0  0  0
23   m   0   0   0  1   0  0  0  0  0  0
24  ak   0   0   0  0   1  0  0  0  0  0
25   m   0   0   0  1   0  0  0  0  0  0
26  as   1   0   0  0   0  0  0  0  0  0
27  as   1   0   0  0   0  0  0  0  0  0
28  as   1   0   0  0   0  0  0  0  0  0
29   m   0   0   0  1   0  0  0  0  0  0
30  as   1   0   0  0   0  0  0  0  0  0
31  as   1   0   0  0   0  0  0  0  0  0
32   m   0   0   0  1   0  0  0  0  0  0
33  as   1   0   0  0   0  0  0  0  0  0
34  as   1   0   0  0   0  0  0  0  0  0
35  ai   0   0   1  0   0  0  0  0  0  0
36  as   1   0   0  0   0  0  0  0  0  0
37   k   0   0   0  0   0  0  0  0  0  0
38  as   1   0   0  0   0  0  0  0  0  0
39  as   1   0   0  0   0  0  0  0  0  0

In [31]:
#top_10
#get whole set of dummy variables, for all the categorical variables
def one_hot_top_x(df, variable, top_x_labels):
    #Function to create the dummy variables for the most frequent labels
    #We can vary the number of most frequent labels that we encode

    for label in top_x_labels:
        df[variable+'_'+label] = np.where(data[variable]==label,1,0)

#read data again
new_data=pd.read_csv('Mercedes_train.csv',usecols=['X1','X2','X3','X4','X5','X6'])
#new_data

#Encode X2 into 10 most frequent categories
one_hot_top_x(new_data,'X2',top_10)
new_data.head()



X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0  v  at  a  d  u  j      0      0      0     0      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0      0     0     1     0   
3  t   n  f  d  x  l      0      0      0     0      0     0     1     0   
4  v   n  f  d  h  d      0      0      0     0      0     0     1     0   

   X2_f  X2_e  
0     0     0  
1     0     0  
2     0     0  
3     0     0  
4     0     0

In [ ]:
##One Hot Encoding of top varaibles
##Advantages
   ###* Straightforward to Implement
   ###* Does not require hrs of variable exploration
   ###* Does not expand massively the feature space(number of columns in the dataset)

##Disadvantages
   ###* Does not add any information that may make the variable more predictive
   ###* Does not keep the Information of the Ignored labels



## 2.1 Label Encoding

In [ ]:
#This is basically applies in a ordinal Categories
#Education
#B.E            2   Labels
#MASTERS        3   Labels
#PhD            4   Labels
#12th           1   Labels


### 2.2 Target Guided ordinal Categories 

In [ ]:
#Suppose for my Categories A o/p is 1, B is 1, C is 0, D is 1, again A is 0, B is 0
#It is repeated several number of times.
#What we do that, we are considering these f1 & o/p as a column.
#Each and every categories of feature f1, I'll tried out to find the mean of this particular values of 1 in o/p column.
#when I'm calculating the mean, It basically means, wherever the A value is 1 and wherever the A value is 0.
#I'm going to considered with respect to those features and I'm going to compute the mean of this value.
# when you see mean, You are trying to find out the number of values for which A is basically 1 and mean is giving for A is o.73, B is 0.6, C is 0.55
#Once you have the mean, what you have to do that, you have assign this rank on the basis of on particular mean.
#Here, I'm talking about oridinal categories, so we are assigning rank on particular mean.
#For A value most of the value is 1 and based on that A have higher mean.

#If A is having highest mean then we will provide highest rank.
#And on the basis of rank, we are providing label like 4,3,2,1,4,3,..... and this value is provided to ML algorithm.

#Because this ordinal, we are providing whole numbers, if it is Nominal then I cannot give this value directly because I don't know ordering in Nominal and I don't know which categorical features is more important.

#For that, I have another features that is MEAN ENCODING.
#If I have nominal Categories, How can we solve?






#Label	f1   O/P    Classification
#4	     A	  1  -> Mean 0.73
#3	     B	  1			 0.60
#2	     C	  0			 0.55
#1	     D	  1          0.33
#4	     A	  0          0.73
#3	     B	  0          0.60


## 1.3 Mean Encoding

In [ ]:
#Suppose I have Feature f1 and o/p

#f1   O/P	Mean
#A	  1 	0.73
#B	  1	    0.60
#C	  0	    0.55
#D	  1     0.33
#A	  0     0.73
#B	  0     0.60 
#what we do that, Instead of converting these into categories, just convert these into a mean values.
#So, This value will be replaced in the place of features(f1) instead of converting into categories like how did I converted in Ordinal.

#Now, You are thinking, where it is basically applicable?

#Let me take sample of pincode
# (Pincodes)f1     O/P  Mean
#  560011	  		1    0.73
#  560012	  		1    0.60
#  560013	  		0    0.55
#  560014	  		1    0.33
#  560015	  		0    0.73
#  560016	  		0    

#Suppose, I have these type of pincodes and I may have city 1000 pincodes and I convert that into One Hot Encoding then I'm creating 999 columns.
#Instead of that, we are trying to convert these into that Integer or float values, based on o/p columns, whether we will see is either 0 or 1 and then tried to find the mean of each categories.

# Suppose these pincodes, 560011 is 0.73, 560012 is 0.60, then All this pincode will replace with this pincodes.

# We are not converting into categorical or dummies variables, that is converting into floating point based on the mean of the output.






## How to Scale Features

In [ ]:
#There are four common methods to perform Feature scaling.
#Standardisation:- Standardisation Replaces the value by their Z scores.
#This redistributes the features with their mean  

## Feature Scaling

In [ ]:
## Why do we required Feature Scaling?
#Whenever we discussed about features scaling means, we discussed about features.


#### There are four common methods to perform Feature Scaling.

##### 1. Standardisation :- where mue(mean) = 0, Standard deviation = 1
##### 2. Mean Normalisation :- Values between -1 and +1, mue(mean) = 0
##### 3. Min-max Normalisation :- between 0 to 1
##### 4. Unit Vector :- range[0,1]


### Examples of Algorithms where Feature Scaling matters

##### 1. k-means uses the Euclidean distance measure here feature scaling matters.
##### 2. K-Nearest-Neighbours also require feature scaling.
##### 3. Principal Component Analysis(PCA): Tries to get the features with maximum variance, here too feature scaling is required.
##### 4. Gradient Descent (Like Linear Regression ): Calculation speed increase as Theta calculation becomes faster after feature scaling.
###### Note: Naive Bayes, Linear Discriminant Analysis, and Tree-Based Models are not affected by feature scaling. In short, any Algorithm which is not Distance based is not affected by Feature Scaling.


### Standardization (Z-Score Normalization) Vs Normalization- Feature Scaling